pkg/hubble/metrics/api/context.go

```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
 	return cs, nil
 }
 
+func parseLabels(s string) (labelsSet, error) {
+	labels := strings.Split(s, ",")
+	for _, label := range labels {
+		if !allowedContextLabels.HasLabel(label) {
+			return labelsSet{}, fmt.Errorf("invalid labelsContext value: %s", label)
+		}
+	}
+	ls := newLabelsSet(labels)
+	return ls, nil
+}
+
 // ParseContextOptions parses a set of options and extracts the context
 // relevant options
 func ParseContextOptions(options Options) (*ContextOptions, error) {
 	o := &ContextOptions{}
+	var err error
 	for key, value := range options {
 		switch strings.ToLower(key) {
 		case "destinationcontext":
-			c, err := parseContext(value)
+			o.Destination, err = parseContext(value)
 			if err != nil {
 				return nil, err
 			}
-			o.Destination = c
 		case "sourcecontext":
-			c, err := parseContext(value)
+			o.Source, err = parseContext(value)
+			if err != nil {
+				return nil, err
+			}
+		case "labelscontext":
+			o.Labels, err = parseLabels(value)
 			if err != nil {
 				return nil, err
 			}
-			o.Source = c
 		}
 	}
 
 	return o, nil
 }
 
+type labelsSet map[string]struct{}
+
+func newLabelsSet(labels []string) labelsSet {
+	m := make(map[string]struct{}, len(labels))
+	for _, label := range labels {
+		m[label] = struct{}{}
+	}
+	return labelsSet(m)
+}
+
+func (ls labelsSet) HasLabel(label string) bool {
+	_, exists := ls[label]
+	return exists
+}
+
+func labelsContext(wantedLabels labelsSet, flow *pb.Flow) (outputLabels []string) {
+	// Iterate over contextLabelsList so that the label order is stable,
+	// otherwise GetLabelNames and GetLabelValues might be mismatched
+	for _, label := range contextLabelsList {
+		if wantedLabels.HasLabel(label) {
+			var labelValue string
+			switch label {
+			case "source_pod":
+				if flow.GetSource() != nil {
+					labelValue = flow.GetSource().PodName
+				}
```

Comment:
nit: protobuf methods are `nil` safe so you can simplify here
```suggestion
				labelValue = flow.GetSource().GetPodName()
```
ditto for the other branches of the `switch`.

Reply:
Ah, I copied this approach from the code we already had in here, so I thought it was necessary. Ill try that.

Reply:
Updated